optimum/habana/transformers/trainer.py

```diff
@@ -736,9 +737,34 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args, use_reentrant: Optio
 
             self.model.gradient_checkpointing_enable(gradient_checkpointing_kwargs=args.gradient_checkpointing_kwargs)
 
+            # Is this necessary ? why is it not in acceleate ?
             # Wrap `_gradient_checkpointing_func` in the model with `transformer_engine` `activation_checkpointing` context.
-            if self.accelerator.state.is_fp8_enabled:
-                FP8ContextWrapper.gradient_checkpointing_wrap(self.model)
+            if self.accelerator.state.mixed_precision == "fp8":
+                import intel_transformer_engine as te
+
+                def _gradient_checkpointing_wrap(func, *args, **kwargs):
+                    """
+                    `_gradient_checkpointing_func` always takes the function to be recomputed as the first argument. The function
+                    below wraps this first argument with `transformer_engine`'s `activation_checkpointing` context.
+                    """
+                    _args = list(args)
+                    _args[0] = te.distributed.activation_checkpointing()(_args[0])
+                    args = tuple(_args)
+
+                    return func(*args, **kwargs)
+
+                if hasattr(self.model, "gradient_checkpointing") and self.model.gradient_checkpointing:
+                    self.model._gradient_checkpointing_func = functools.partial(
+                        _gradient_checkpointing_wrap, self.model._gradient_checkpointing_func
+                    )
+                    return
+
+                for module in self.model.modules():
+                    if hasattr(module, "gradient_checkpointing") and module.gradient_checkpointing:
+                        module._gradient_checkpointing_func = functools.partial(
+                            _gradient_checkpointing_wrap, module._gradient_checkpointing_func
+                        )
```

Comment:
I moved this from GaudiAccelerator's fp8 implementation, for now it resides here but will be ported to upstream accelerate later.

Reply:
if this is generalizable it could be moved to accelerate, but not sure if it's still relevant.